File: go/meta_schemas_test.go
package jsonstructure

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

func loadSchema(t *testing.T, path string) map[string]interface{} {
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("Failed to read schema file %s: %v", path, err)
	}

	var schema map[string]interface{}
	if err := json.Unmarshal(data, &schema); err != nil {
		t.Fatalf("Failed to parse schema file %s: %v", path, err)
	}

	return schema
}

func TestMetaSchemas(t *testing.T) {
	// Get the path to the meta directory
	metaDir := filepath.Join("..", "meta")

	// Load all three metaschemas
	coreSchema := loadSchema(t, filepath.Join(metaDir, "core", "v0", "index.json"))
	extendedSchema := loadSchema(t, filepath.Join(metaDir, "extended", "v0", "index.json"))
	validationSchema := loadSchema(t, filepath.Join(metaDir, "validation", "v0", "index.json"))

	// Build external schemas map for import resolution
	externalSchemas := map[string]interface{}{
		coreSchema["$id"].(string):       coreSchema,
		extendedSchema["$id"].(string):   extendedSchema,
		validationSchema["$id"].(string): validationSchema,
	}

	t.Run("CoreMetaschema", func(t *testing.T) {
		validator := NewSchemaValidator(&SchemaValidatorOptions{
			AllowDollar:     true,
			AllowImport:     true,
			ExternalSchemas: externalSchemas,
		})

		// Reload fresh copy
		schema := loadSchema(t, filepath.Join(metaDir, "core", "v0", "index.json"))
		result := validator.Validate(schema)

		if !result.IsValid {
			for _, err := range result.Errors {
				t.Errorf("Validation error: %s at %s", err.Message, err.Path)
			}
		}
	})

	t.Run("ExtendedMetaschema", func(t *testing.T) {
		validator := NewSchemaValidator(&SchemaValidatorOptions{
			AllowDollar:     true,
			AllowImport:     true,
			ExternalSchemas: externalSchemas,
		})

		// Reload fresh copy
		schema := loadSchema(t, filepath.Join(metaDir, "extended", "v0", "index.json"))
		result := validator.Validate(schema)

		if !result.IsValid {
			for _, err := range result.Errors {
				t.Errorf("Validation error: %s at %s", err.Message, err.Path)
			}
		}
	})

	t.Run("ValidationMetaschema", func(t *testing.T) {
		validator := NewSchemaValidator(&SchemaValidatorOptions{
			AllowDollar:     true,
			AllowImport:     true,
			ExternalSchemas: externalSchemas,
		})

		// Reload fresh copy
		schema := loadSchema(t, filepath.Join(metaDir, "validation", "v0", "index.json"))
		result := validator.Validate(schema)

		if !result.IsValid {
			for _, err := range result.Errors {
				t.Errorf("Validation error: %s at %s", err.Message, err.Path)
			}
		}
	})

	t.Run("CoreMetaschemaHasDefinitions", func(t *testing.T) {
		defs, ok := coreSchema["definitions"].(map[string]interface{})
		if !ok {
			t.Fatal("Core metaschema missing definitions")
		}

		expectedDefs := []string{"SchemaDocument", "ObjectType", "Property"}
		for _, name := range expectedDefs {
			if _, exists := defs[name]; !exists {
				t.Errorf("Core metaschema missing expected definition: %s", name)
			}
		}
	})

	t.Run("ExtendedMetaschemaImportsCore", func(t *testing.T) {
		// Reload fresh copy to check original $import
		schema := loadSchema(t, filepath.Join(metaDir, "extended", "v0", "index.json"))
		importURI, ok := schema["$import"].(string)
		if !ok {
			t.Fatal("Extended metaschema missing $import")
		}
		if importURI != "https://json-structure.org/meta/core/v0/#" {
			t.Errorf("Expected $import to be core URI, got: %s", importURI)
		}
	})

	t.Run("ValidationMetaschemaImportsExtended", func(t *testing.T) {
		// Reload fresh copy to check original $import
		schema := loadSchema(t, filepath.Join(metaDir, "validation", "v0", "index.json"))
		importURI, ok := schema["$import"].(string)
		if !ok {
			t.Fatal("Validation metaschema missing $import")
		}
		if importURI != "https://json-structure.org/meta/extended/v0/#" {
			t.Errorf("Expected $import to be extended URI, got: %s", importURI)
		}
	})
}

File: typescript/tests/meta-schemas.test.ts
import { describe, it, expect, beforeAll } from 'vitest';
import { SchemaValidator } from '../src/schema-validator';
import { JsonObject, JsonValue } from '../src/types';
import * as fs from 'fs';
import * as path from 'path';

describe('Meta Schema Validation', () => {
  let coreSchema: JsonObject;
  let extendedSchema: JsonObject;
  let validationSchema: JsonObject;
  let externalSchemas: Map<string, JsonValue>;

  beforeAll(() => {
    const metaDir = path.join(__dirname, '..', '..', 'meta');
    
    // Load all three metaschemas
    coreSchema = JSON.parse(
      fs.readFileSync(path.join(metaDir, 'core', 'v0', 'index.json'), 'utf-8')
    );
    extendedSchema = JSON.parse(
      fs.readFileSync(path.join(metaDir, 'extended', 'v0', 'index.json'), 'utf-8')
    );
    validationSchema = JSON.parse(
      fs.readFileSync(path.join(metaDir, 'validation', 'v0', 'index.json'), 'utf-8')
    );

    // Build external schemas map for import resolution
    externalSchemas = new Map<string, JsonValue>();
    externalSchemas.set(coreSchema.$id as string, coreSchema);
    externalSchemas.set(extendedSchema.$id as string, extendedSchema);
    externalSchemas.set(validationSchema.$id as string, validationSchema);
  });

  describe('core metaschema', () => {
    it('should validate without errors', () => {
      const validator = new SchemaValidator({
        allowDollar: true,
        allowImport: true,
        externalSchemas,
      });
      const result = validator.validate(coreSchema);
      
      if (!result.isValid) {
        console.log('Core metaschema validation errors:', result.errors);
      }
      
      expect(result.isValid).toBe(true);
      expect(result.errors).toHaveLength(0);
    });

    it('should have expected definitions', () => {
      expect(coreSchema.definitions).toBeDefined();
      expect((coreSchema.definitions as JsonObject).SchemaDocument).toBeDefined();
      expect((coreSchema.definitions as JsonObject).ObjectType).toBeDefined();
      expect((coreSchema.definitions as JsonObject).Property).toBeDefined();
    });
  });

  describe('extended metaschema', () => {
    it('should validate without errors', () => {
      // Need to reload to get fresh copy since import processing modifies the schema
      const metaDir = path.join(__dirname, '..', '..', 'meta');
      const freshExtended = JSON.parse(
        fs.readFileSync(path.join(metaDir, 'extended', 'v0', 'index.json'), 'utf-8')
      );
      
      const validator = new SchemaValidator({
        allowDollar: true,
        allowImport: true,
        externalSchemas,
      });
      const result = validator.validate(freshExtended);
      
      if (!result.isValid) {
        console.log('Extended metaschema validation errors:', result.errors);
      }
      
      expect(result.isValid).toBe(true);
      expect(result.errors).toHaveLength(0);
    });

    it('should import from core', () => {
      // Reload fresh copy to check the original $import
      const metaDir = path.join(__dirname, '..', '..', 'meta');
      const freshExtended = JSON.parse(
        fs.readFileSync(path.join(metaDir, 'extended', 'v0', 'index.json'), 'utf-8')
      );
      expect(freshExtended.$import).toBe('https://json-structure.org/meta/core/v0/#');
    });
  });

  describe('validation metaschema', () => {
    it('should validate without errors', () => {
      // Need to reload to get fresh copy since import processing modifies the schema
      const metaDir = path.join(__dirname, '..', '..', 'meta');
      const freshValidation = JSON.parse(
        fs.readFileSync(path.join(metaDir, 'validation', 'v0', 'index.json'), 'utf-8')
      );
      
      const validator = new SchemaValidator({
        allowDollar: true,
        allowImport: true,
        externalSchemas,
      });
      const result = validator.validate(freshValidation);
      
      if (!result.isValid) {
        console.log('Validation metaschema validation errors:', result.errors);
      }
      
      expect(result.isValid).toBe(true);
      expect(result.errors).toHaveLength(0);
    });

    it('should import from extended', () => {
      // Reload fresh copy to check the original $import
      const metaDir = path.join(__dirname, '..', '..', 'meta');
      const freshValidation = JSON.parse(
        fs.readFileSync(path.join(metaDir, 'validation', 'v0', 'index.json'), 'utf-8')
      );
      expect(freshValidation.$import).toBe('https://json-structure.org/meta/extended/v0/#');
    });
  });
});
